File: python/015_three_sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        for a in range(len(nums)):
            b, c = a+1, len(nums)-1
            while b < c:
                if nums[a] + nums[b] + nums[c] < 0:
                    b += 1
                elif nums[a] + nums[b] + nums[c] > 0:
                    c -= 1
                else:
                    if [nums[a], nums[b], nums[c]] not in result:
                        result.append([nums[a], nums[b], nums[c]])
                    b += 1
        return result


if __name__ == "__main__":
    num_list = [-1, 0, 1, 2, -1, -4]
    s = Solution()
    print(s.threeSum(num_list))

File: cpp/015_three_sum.cpp
#include "header.h"
using namespace std;

// two pointer
class Solution {
public:
    vector< vector<int> > threeSum(vector<int>& nums) {
        if (nums.size() <= 2) return {};
        vector< vector<int> > ans;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for (int i=0; i<n-2; ++i) {
            if (nums[i] > 0) break;
            if (i>0 && nums[i]==nums[i-1]) continue;
            int l = i+1;
            int r = n-1;
            while (l<r) {
                if (nums[i] + nums[l] + nums[r] == 0){
                    ans.push_back({nums[i], nums[l++], nums[r--]});
                    while (l<r && nums[l] == nums[l-1]) ++l;
                    while (l<r && nums[r] == nums[r+1]) --r;
                }
                else if (nums[i] + nums[l] + nums[r] < 0) ++l;
                else --r;
            }
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    cout << s.threeSum(nums)[0][0] << endl;
}